### PR TITLE
[MPS] faster norms

### PR DESCRIPTION
**Pull request by Isalia20**
Routes `norm(1/2, dim=-1)` through the efficient `sum_reduction_inner` kernel via abs/square load + sqrt finalize, instead of the slow per-output-threadgroup norm kernel with per-element `pow`. 
Example:

```
import torch
import torch.nn.functional as F

x = torch.randn(4096, 4096, device="mps", dtype=torch.bfloat16)

x.norm(2, dim=-1)          # L2 -> norm_l2_reduction_inner kernel
x.norm(1, dim=-1)          # L1 -> norm_l1_reduction_inner kernel
F.normalize(x, dim=-1)
```

Perf:

### fp32
| shape | old (µs) | new (µs) | speedup |
|---|--:|--:|--:|
| (32,768) | 6.9 | 3.8 | **1.8x** |
| (128,1024) | 17.1 | 4.2 | **4.1x** |
| (512,4096) | 171.9 | 20.9 | **8.2x** |
| (2048,4096) | 676.8 | 114.5 | **5.9x** |
| (4096,4096) | 1348.2 | 238.4 | **5.7x** |
| (16384,1024) | 1601.8 | 236.2 | **6.8x** |
| (1024,8192) | 666.3 | 112.0 | **5.9x** |
| (65536,256) | 1707.3 | 229.9 | **7.4x** |
| (8,32,768) | 27.5 | 4.2 | **6.6x** |
| (32,128,128) | 73.1 | 7.4 | **9.9x** |

### bf16
| shape | old (µs) | new (µs) | speedup |
|---|--:|--:|--:|
| (32,768) | 6.9 | 3.6 | **1.9x** |
| (128,1024) | 17.0 | 3.7 | **4.6x** |
| (512,4096) | 177.5 | 10.3 | **17.2x** |
| (2048,4096) | 715.7 | 40.0 | **17.9x** |
| (4096,4096) | 1402.8 | 117.6 | **11.9x** |
| (16384,1024) | 1650.8 | 112.5 | **14.7x** |
| (1024,8192) | 678.0 | 42.4 | **16.0x** |
| (65536,256) | 1790.2 | 113.2 | **15.8x** |
| (8,32,768) | 27.4 | 3.9 | **7.0x** |
| (32,128,128) | 74.5 | 7.5 | **10.0x** |

### fp16
| shape | old (µs) | new (µs) | speedup |
|---|--:|--:|--:|
| (32,768) | 6.8 | 3.8 | **1.8x** |
| (128,1024) | 17.0 | 3.9 | **4.3x** |
| (512,4096) | 172.6 | 10.2 | **16.9x** |
| (2048,4096) | 678.6 | 40.4 | **16.8x** |
| (4096,4096) | 1352.9 | 112.8 | **12.0x** |
| (16384,1024) | 1604.2 | 107.6 | **14.9x** |
| (1024,8192) | 670.2 | 42.5 | **15.8x** |
| (65536,256) | 1711.6 | 108.1 | **15.8x** |
| (8,32,768) | 27.5 | 4.0 | **6.8x** |
| (32,128,128) | 73.2 | 7.7 | **9.5x** |

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/ReduceOps.metal [hunk 1]
```diff
@@ -230,6 +244,29 @@ inline uint load_val(TI v) {
...
+    typename TI,
+    ::metal::enable_if_t<MODE == LOAD_SQUARE, bool> = true>
+inline opmath_t<TI> load_val(TI v) {
+  auto r = static_cast<opmath_t<TI>>(v);
+  return r * r;
+}
+
+template <FinalizeOp FINAL, typename T>
+inline T finalize_val(T v) {
+  return FINAL == FINAL_SQRT ? static_cast<T>(::precise::sqrt(v)) : v;
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/kernels/ReduceOps.metal]: Why not use `enable_if_t` to have two overloads rather than this conditional?

File: aten/src/ATen/native/mps/operations/ReduceOps.mm [hunk 2]
```diff
@@ -277,6 +277,46 @@ static void norm_kernel_mps(TensorIterator& iter, const Scalar& p_scalar) {
...
+      }
+    }
+    if (num_reduced == 1 && reduced_dim == input.dim() - 1) {
+      uint32_t N = input.size(input.dim() - 1);
+      uint32_t M = input.numel() / N;
+      auto kernel_name = fmt::format("norm_{}_reduction_inner_{}_{}",
+                                     p == 2.0 ? "l2" : "l1",
+                                     scalarToMetalTypeString(input),
+                                     scalarToMetalTypeString(output));
+      constexpr uint32_t TG_SIZE = 256;
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/operations/ReduceOps.mm:300]: Why not use all 1024?

File: aten/src/ATen/native/mps/operations/ReduceOps.mm [hunk 3]
```diff
@@ -277,6 +277,46 @@ static void norm_kernel_mps(TensorIterator& iter, const Scalar& p_scalar) {
...
+      constexpr uint32_t rows_per_tg = TG_SIZE / 32;
+      const auto num_tgs = c10::metal::ceil_div(M, rows_per_tg);
+      MPSStream* stream = getCurrentMPSStream();
+      dispatch_sync_with_rethrow(stream->queue(), ^() {
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/ReduceOps.mm]: Nit (I sort of dislike this syntax, but imo better than explicit `return` line right after the call)
```suggestion
      return dispatch_sync_with_rethrow(stream->queue(), ^() {
```

File: aten/src/ATen/native/mps/operations/ReduceOps.mm [hunk 4]
```diff
@@ -277,6 +277,46 @@ static void norm_kernel_mps(TensorIterator& iter, const Scalar& p_scalar) {
...
+        @autoreleasepool {
+          id<MTLComputeCommandEncoder> ce = stream->commandEncoder();
+          auto ps = lib.getPipelineStateForFunc(kernel_name);
+          getMPSProfiler().beginProfileKernel(ps, "norm_reduction_inner", {input});
+          const float divisor = 0.0f;
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/ReduceOps.mm]: Why not pass it as 0.0f directly?

File: aten/src/ATen/native/mps/kernels/ReduceOps.metal [hunk 5]
```diff
@@ -149,7 +155,9 @@ kernel void norm(
       }
     }
 
-    if (p != 0 && p != 1 && p != INFINITY && p != -INFINITY) {
+    if (p == 2) {
+      output_val = static_cast<TA>(::precise::sqrt(output_val));
+    } else if (p != 0 && p != 1 && p != INFINITY && p != -INFINITY) {
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/ReduceOps.metal]: Nit: why not do it inside `if (p != 0 ...)?

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/mps/operations/ReduceOps.mm (hunk 2 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/ReduceOps.mm:300]: simply a mirror from below which uses 256, I benchmarked with 1024 as well and it gives the same speedup

File: aten/src/ATen/native/mps/operations/ReduceOps.mm (hunk 3 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/ReduceOps.mm]: fixed

File: aten/src/ATen/native/mps/operations/ReduceOps.mm (hunk 4 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/ReduceOps.mm]: fixed

File: aten/src/ATen/native/mps/kernels/ReduceOps.metal (hunk 1 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/kernels/ReduceOps.metal]: fixed

File: aten/src/ATen/native/mps/kernels/ReduceOps.metal (hunk 5 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/kernels/ReduceOps.metal]: if you meant ternary op inside for p==2 then fixed
